sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive
+
+import java.util.concurrent.atomic.AtomicLong
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.SQLContext
+import org.apache.spark.sql.catalyst.expressions.{NamedExpression, Attribute, SortOrder}
+import org.apache.spark.sql.catalyst.optimizer.ProjectCollapsing
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.{Rule, RuleExecutor}
+import org.apache.spark.sql.catalyst.util.sequenceOption
+
+class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Logging {
+  def toSQL: Option[String] = {
+    val canonicalizedPlan = Canonicalizer.execute(logicalPlan)
+
+    logDebug(
+      s"""Building SQL query string from given logical plan:
+         |
+         |# Original logical plan:
+         |${logicalPlan.treeString}
+         |# Canonicalized logical plan:
+         |${canonicalizedPlan.treeString}
+       """.stripMargin)
+
+    toSQL(canonicalizedPlan)
+  }
+
+  private def toSQL(node: LogicalPlan): Option[String] = node match {
+    case plan @ Project(list, child) =>
+      for {
+        listSQL <- sequenceOption(list.map(_.sql))
+        childSQL <- toSQL(child)
```

Comment:
Good catch! But `getOrElse` isn't applicable here, because `child` may be some complex query plan other than `OneRowRelation` but we don't support (e.g. a `LocalRelation` built from local Scala `Seq`). I'd add a branch for `OneRowRelation` that returns `Some("")`.